sdk/core/Azure.Core/src/Shared/HttpMessageSanitizer.cs

```diff
@@ -0,0 +1,134 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Text;
+using System.Linq;
+
+namespace Azure.Core
+{
+    internal class HttpMessageSanitizer
+    {
+        private readonly bool _logAllHeaders;
+        private readonly bool _logFullQueries;
+        private readonly string[] _allowedQueryParameters;
+        private readonly string _redactedPlaceholder;
+        private readonly HashSet<string> _allowedHeaders;
+        private const string LogAllValue = "*";
+
+        public HttpMessageSanitizer(string[] allowedQueryParameters, string[] allowedHeaders, string redactedPlaceholder = "REDACTED")
+        {
+            _logAllHeaders = allowedHeaders.Contains(LogAllValue);
+            _logFullQueries = allowedQueryParameters.Contains(LogAllValue);
+
+            _allowedQueryParameters = allowedQueryParameters;
+            _redactedPlaceholder = redactedPlaceholder;
+            _allowedHeaders = new HashSet<string>(allowedHeaders, StringComparer.InvariantCultureIgnoreCase);
+        }
+
+        public string SanitizeHeader(string name, string value)
+        {
+            if (_logAllHeaders || _allowedHeaders.Contains(name))
```

Comment:
As often as this gets called, should we index header names (dictionary)?

Reply:
_allowedHeaders is a hashset.